hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmCriteriaNodeBuilder.java

```diff
@@ -1827,8 +1827,11 @@ private static <T> BindableType<T> resolveInferredParameterType(
 			if ( typeInferenceSource instanceof BindableType ) {
 				return (BindableType<T>) typeInferenceSource;
 			}
-			if ( typeInferenceSource.getNodeType() != null ) {
-				return (BindableType<T>) typeInferenceSource.getNodeType();
+			final SqmExpressible<?> nodeType = typeInferenceSource instanceof SqmPath<?> ?
+					( (SqmPath<?>) typeInferenceSource ).getResolvedModel().getSqmType() :
+					typeInferenceSource.getNodeType();
+			if ( nodeType != null ) {
+				return (BindableType<T>) nodeType;
```

Comment:
This was done to fix an existing test with a generic MappedSuperclass comparison predicate that failed type-checking. If the change in https://github.com/hibernate/hibernate-orm/pull/7590 will be approved, I believe it could be simplified to use `typeInferenceSource.getExpressible()`.

Reply:
@mbladel Just to remind you this could be simplified now.